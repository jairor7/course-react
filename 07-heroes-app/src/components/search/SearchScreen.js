import React, { useEffect, useState } from "react";
import { useNavigate, useSearchParams } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { getHeroByName } from "../../selectors/heroes";
import { HeroCards } from "../heros/HeroCards";
export const SearchScreen = () => {
  const [heroesFilter, setHeroesFilter] = useState([]);
  const [searchParams, setSearchParams] = useSearchParams();
  const [{ searchText }, handleInputChange] = useForm({
    searchText: searchParams.get("hero") || "",
  });
  const navigate = useNavigate();

  useEffect(() => {
    const heroesSearched = searchParams.get("hero");
    if (!!heroesSearched)
      setHeroesFilter(getHeroByName(heroesSearched));
    else setHeroesFilter([]);
  }, [searchParams]);

  const handleSubmit = (e) => {
    e.preventDefault();
    // Esta linea tambien funciona pero no con las pruebas
    setSearchParams({ hero: searchText });
    // navigate(`?hero=${searchText}`);
  };

  return (
    <>
      <h1>Search Screen</h1>
      <hr />

      <div className="row">
        <div className="col-5">
          <h5>Search</h5>
          <hr />
          <form onSubmit={handleSubmit}>
            <div className="row">
              <div className="col-8">
                <input
                  className="form-control"
                  type="text"
                  name="searchText"
                  placeholder="Search"
                  value={searchText}
                  autoComplete={"off"}
                  onChange={handleInputChange}
                />
              </div>
              <div className="col-4">
                <button
                  className="btn btn-outline-primary"
                  type="submit"
                >
                  Buscar...
                </button>
              </div>
            </div>
          </form>
        </div>
        <div className="col-7">
          <h5>Result</h5>
          <hr />
          {!searchParams.get("hero") ? (
            <div className="alert alert-info">Buscar un h√©roe</div>
          ) : (
            heroesFilter.length <= 0 && (
              <div className="alert alert-danger">
                No hay resultados: {searchParams.get("hero")}
              </div>
            )
          )}

          {heroesFilter.map((hero) => (
            <HeroCards key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </>
  );
};
