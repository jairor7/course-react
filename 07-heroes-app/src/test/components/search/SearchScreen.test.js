import { SearchScreen } from "../../../components/search/SearchScreen";
import { mount } from "enzyme";
import { MemoryRouter, useNavigate } from "react-router-dom";

// Esta prueba funciona pero al mockear el useSearchParams las demas pruebas fallan

const mockUseSearchParams = jest.fn();
jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useSearchParams: () => [
    jest.requireActual("react-router-dom").useSearchParams()[0],
    mockUseSearchParams,
  ],
}));

// const mockNavigate = jest.fn();

// jest.mock("react-router-dom", () => ({
//   ...jest.requireActual("react-router-dom"),
//   useNavigate: () => mockNavigate,
// }));

describe("Pruebas en el <SearchScreen />", () => {
  test("Debe mostrar el componente correctamente", () => {
    const wrapper = mount(
      <MemoryRouter
        children={<SearchScreen />}
        initialEntries={["/search"]}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert.alert-info").text().trim()).toBe(
      "Buscar un hÃ©roe"
    );
  });

  test("Debe de mostrar a Batman y el input con el valor del queryString", () => {
    const wrapper = mount(
      <MemoryRouter
        children={<SearchScreen />}
        initialEntries={["/search?hero=batman"]}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("input").prop("value")).toBe("batman");
  });

  test("Debe de mostrar un error, si no se encuentra el hero", () => {
    const wrapper = mount(
      <MemoryRouter
        children={<SearchScreen />}
        initialEntries={["/search?hero=123"]}
      />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(".alert.alert-danger").text().trim()).toBe(
      "No hay resultados: 123"
    );
  });

  test("Debe de cambiar la url para realizar un nuevo searchParams", () => {
    const wrapper = mount(
      <MemoryRouter
        children={<SearchScreen />}
        initialEntries={["/search?hero=su"]}
      />
    );
    wrapper.find("input").simulate("change", {
      target: { name: "searchText", value: "batman" },
    });

    wrapper.find("form").prop("onSubmit")({
      preventDefault: () => {},
    });

    expect(mockUseSearchParams).toHaveBeenCalledWith({
      hero: "batman",
    });

    // expect(mockNavigate).toHaveBeenCalledWith("?hero=batman");
  });
});
