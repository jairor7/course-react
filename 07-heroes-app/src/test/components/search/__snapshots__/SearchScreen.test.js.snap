// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <SearchScreen /> Debe de mostrar a Batman y el input con el valor del queryString 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="row"
        >
          <div
            className="col-8"
          >
            <input
              autoComplete="off"
              className="form-control"
              name="searchText"
              onChange={[Function]}
              placeholder="Search"
              type="text"
              value="batman"
            />
          </div>
          <div
            className="col-4"
          >
            <button
              className="btn btn-outline-primary"
              type="submit"
            >
              Buscar...
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h5>
        Result
      </h5>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-5"
            >
              <img
                alt="Batman"
                className="card-img-top"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-7"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="text-muted"
                >
                  Detective Comics #27
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Mas información
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en el <SearchScreen /> Debe de mostrar un error, si no se encuentra el hero 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="row"
        >
          <div
            className="col-8"
          >
            <input
              autoComplete="off"
              className="form-control"
              name="searchText"
              onChange={[Function]}
              placeholder="Search"
              type="text"
              value="123"
            />
          </div>
          <div
            className="col-4"
          >
            <button
              className="btn btn-outline-primary"
              type="submit"
            >
              Buscar...
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h5>
        Result
      </h5>
      <hr />
      <div
        className="alert alert-danger"
      >
        No hay resultados: 
        123
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en el <SearchScreen /> Debe mostrar el componente correctamente 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <div
          className="row"
        >
          <div
            className="col-8"
          >
            <input
              autoComplete="off"
              className="form-control"
              name="searchText"
              onChange={[Function]}
              placeholder="Search"
              type="text"
              value=""
            />
          </div>
          <div
            className="col-4"
          >
            <button
              className="btn btn-outline-primary"
              type="submit"
            >
              Buscar...
            </button>
          </div>
        </div>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h5>
        Result
      </h5>
      <hr />
      <div
        className="alert alert-info"
      >
        Buscar un héroe
      </div>
    </div>
  </div>,
]
`;
